Update timescaledb_information views (#495)
timescale/docs.timescale.com-content: api.md

| `replication_factor` | Replication factor for a distributed hypertable|
| `data_nodes` | Nodes on which hypertable is distributed|
| `tablespaces` |Tablespaces attached to the hypertable |

#### Sample Usage [](timescaledb_information-hypertables-examples)

Get information about a hypertable.

```sql
CREATE TABLE dist_table(time timestamptz, device int, temp float);
SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);

SELECT * FROM timescaledb_information.hypertables
  WHERE table_name = 'dist_table';

-[ RECORD 1 ]-------+-----------
table_schema        | public
table_name          | dist_table
owner               | postgres 
num_dimensions      | 2
num_chunks          | 3
compression_enabled | f
is_distributed      | t
replication_factor  | 2
data_nodes          | {node_1, node_2}
tablespaces         | 

```

## timescaledb_information.dimensions [](timescaledb_information-dimensions)

Get metadata about the dimensions of hypertables, returning one row of metadata for each dimension of a hypertable.  For a time-and-space-partitioned hypertable, for example, two rows of metadata will be returned for the hypertable.

> If the dimension is time based, the dimension column has either an integer datatye or a time-based datatype. The `time_interval` column is defined for hypertables that use time-based datatypes (e.g., timestamptz, timestamp, date).  Alternatively, for hypertables that use integer datatypes (e.g., bigint, integer), the `integer_interval` and `integer_now_func` columns are defined.

Reviewer: Doesn't that first sentence contradict the 2nd and 3rd sentence. Did you mean to say open dimension in the first sentence?

Author: I think it is confusing. Let me reword that.

> For space based dimensions, metadata is returned that specifies their number of `num_partitions`. The `time_interval` and `integer_interval` columns are not applicable for space based dimensions.

Reviewer: should this be closed dimensions? not sure we defined those terms or use them in other parts

Author: No we don't define open/closed dimensions anywhere. That's why we moved to Time/Space terminology.

Get information about a hypertable.

```sql
CREATE TABLE dist_table(time timestamptz, device int, temp float);
SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);

SELECT * FROM timescaledb_information.hypertables
  WHERE table_name = 'dist_table';

-[ RECORD 1 ]-------+-----------
table_schema        | public
table_name          | dist_table
owner               | postgres 
num_dimensions      | 2
num_chunks          | 3
compression_enabled | f
is_distributed      | t
replication_factor  | 2
data_nodes          | {node_1, node_2}
tablespaces         | 

```

## timescaledb_information.dimensions [](timescaledb_information-dimensions)

Get metadata about the dimensions of hypertables, returning one row of metadata 
for each dimension of a hypertable.  For a time-and-space-partitioned 
hypertable, for example, two rows of metadata will be returned for the 
hypertable.

A time-based dimension column has either an integer datatype (e.g bigint, integer)
or a time related datatype (e.g. timestamptz, timestamp, date).
The `time_interval` column is defined for hypertables that use time datatypes 

> like timestamptz, timestamp, date. 

Reviewer: since those are the only supported datatypes for time dimensions the like is slightly out of place

Author: updated. Hope this reads better.

> Alternatively, for hypertables that use integer datatypes (e.g., bigint, 

Reviewer: probably makes sense to list all datatypes here since there are only 3 smallint,integer,bigint

SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);

SELECT * FROM timescaledb_information.hypertables
WHERE table_name = 'dist_table';
 table_schema | table_name |    owner    | num_dimensions | num_chunks | compression_enabled | is_distributed | replication_factor |              data_nodes               | tablespaces
--------------+------------+-------------+----------------+------------+---------------------+----------------+--------------------+---------------------------------------+-------------
 public       | dist_table | postgres    |              2 |          3 | f                   | t              |                  2 | {node_1,node_2,node_3}     |
(1 row)
```

## timescaledb_information.dimensions [](timescaledb_information-dimensions)

Get metadata about the dimensions of hypertables.
The primary dimension of a hypertable is time based.
The time_interval column is set for hypertables that use date/time datatypes, like timestamp, data etc., for the primary dimension.
The integer_interval and integer_now_func columns are set for hypertables that use integer datatypes, like bigint, integer etc., for the primary dimension.

> Secondary dimensions are space based and have num_partitions set. The time_interval and integer_interval columns are not applicable to secondary dimensions.

Author: Modified the text to reflect that possibility. Added an example with time+time dimensions.

(2 rows)
```

## timescaledb_information.chunks [](timescaledb_information-chunks)

Get metadata about the chunks of hypertables.
This view shows metadata for the chunk's primary dimension.
range_start_integer and range_end_integer are set if the type of the primary dimension type is integer based.
range_start and range_end are set if the primary dimension is of time datatype.

#### Available Columns [](timescaledb_information-chunks-available-columns)

|Name|Description|
|---|---|
| `hypertable_schema` | Schema name of the hypertable |
| `hypertable_name` | Table name of the hypertable |
| `chunk_schema` | Schema name of the chunk |
| `chunk_name` | Name of the chunk |
| `primary_dimension` | Name of the column that is the primary dimension|
| `primary_dimension_type` | Type of the column that is the primary dimension|
| `range_start` | Start of the range for the chunk's dimension |
| `range_end` | End of the range for the chunk's dimension |
| `range_start_integer` | Start of the range for the chunk's dimension, if the dimension type is integer based |

> | `range_end_integer` | End of the range for the chunk's dimension, if the dimension type is integer based. |

Reviewer 2: ```suggestion
| `range_end_integer` | End of the range for the chunk's dimension, if the dimension type is integer based |
```

(2 rows)
```

## timescaledb_information.hypertables [](timescaledb_information-hypertables)

Get metadata information about hypertables.

#### Available Columns [](timescaledb_information-hypertables-available-columns)

|Name|Description|
|---|---|
| `table_schema` | Schema name of the hypertable |
| `table_name` | Table name of the hypertable |
| `table_owner` | Owner of the hypertable |
| `num_dimensions` | Number of dimensions |
| `num_chunks` | Number of chunks |
| `compression_enabled` |Is compression enabled on the hypertable?|
| `is_distributed` | Is the hypertable distributed?|
| `replication_factor` | Replication factor for a distributed hypertable|
| `data_nodes` | Nodes on which hypertable is distributed|
| `tablespaces` |Tablespaces attached to the hypertable |

#### Sample Usage [](timescaledb_information-hypertables-examples)

Get information about a hypertable.

```sql
CREATE TABLE dist_table(time timestamptz, device int, temp float);
SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);

SELECT * FROM timescaledb_information.hypertables
  WHERE table_name = 'dist_table';

>  table_schema | table_name |    owner    | num_dimensions | num_chunks | compression_enabled | is_distributed | replication_factor |              data_nodes               | tablespaces

Reviewer 2: Maybe switch to `\x` for extended views?

Author: fixed.

| `hypertable_schema` | Schema name of the hypertable |
| `hypertable_name` | Table name of the hypertable |
| `dimension_number` | Dimension number of the hypertable, starting from 1 |
| `column_name` | Name of the column used to create this dimension |
| `column_type` | Type of the column used to create this dimension|
| `dimension_type` |Is this time based or space based dimension?|
| `time_interval` | Time interval for primary dimension if the column type is based on Postgres time datatypes |
| `integer_interval` | Integer interval for primary dimension if the column type is an integer datatype |
| `integer_now_func` | integer_now function for primary dimension if the column type is integer based datatype|
| `num_partitions` | Number of partitions for the dimension |

#### Sample Usage [](timescaledb_information-dimensions-examples)

Get information about the dimensions of hypertables.

```sql
CREATE TABLE dist_table(time timestamptz, device int, temp float);
SELECT create_hypertable('dist_table', 'time',  'device', chunk_time_interval=> INTERVAL '7 days', number_partitions=>3);

SELECT * from timescaledb_information.dimensions
  ORDER BY hypertable_name, dimension_number;

>  hypertable_schema | hypertable_name | dimension_number | column_name |       column_type        | dimension_type | time_interval | integer_interval | integer_now_func | num_partitions

Reviewer 2: Maybe switch to extended view `\x` for these longer outputs?

Author: fixed.

This view shows metadata for the chunk's primary dimension.
range_start_integer and range_end_integer are set if the type of the primary dimension type is integer based.
range_start and range_end are set if the primary dimension is of time datatype.

#### Available Columns [](timescaledb_information-chunks-available-columns)

|Name|Description|
|---|---|
| `hypertable_schema` | Schema name of the hypertable |
| `hypertable_name` | Table name of the hypertable |
| `chunk_schema` | Schema name of the chunk |
| `chunk_name` | Name of the chunk |
| `primary_dimension` | Name of the column that is the primary dimension|
| `primary_dimension_type` | Type of the column that is the primary dimension|
| `range_start` | Start of the range for the chunk's dimension |
| `range_end` | End of the range for the chunk's dimension |
| `range_start_integer` | Start of the range for the chunk's dimension, if the dimension type is integer based |
| `range_end_integer` | End of the range for the chunk's dimension, if the dimension type is integer based. |
| `is_compressed` | Is the data in the chunk compressed?|
| `chunk_tablespace` | Tablespace used by the chunk|

> | `data_nodes` | Nodes on which the chunk is replicated. This is applicable only to chunks for distributed hypertables. |

Reviewer 2: ```suggestion
| `data_nodes` | Nodes on which the chunk is replicated. This is applicable only to chunks for distributed hypertables |
```

#### Sample Usage [](timescaledb_information-chunks-examples)

Get information about the chunks of a hypertable.

```sql
CREATE TABLESPACE tablespace1 location '/usr/local/pgsql/data1';
  
CREATE TABLE hyper_int (a_col integer, b_col integer, c integer);
SELECT table_name from create_hypertable('hyper_int', 'a_col', chunk_time_interval=> 10);
CREATE OR REPLACE FUNCTION integer_now_hyper_int() returns int LANGUAGE SQL STABLE as $$ SELECT coalesce(max(a_col), 0) FROM hyper_int $$;
SELECT set_integer_now_func('hyper_int', 'integer_now_hyper_int');

INSERT INTO hyper_int SELECT generate_series(1,5,1), 10, 50;

SELECT attach_tablespace('tablespace1', 'hyper_int');
INSERT INTO hyper_int VALUES( 25 , 14 , 20), ( 25, 15, 20), (25, 16, 20);

SELECT * FROM timescaledb_information.chunks WHERE hypertable_name = 'hyper_int';

>  hypertable_schema | hypertable_name |     chunk_schema      |    chunk_name     | primary_di

Reviewer 2: Maybe switch to extended view `\x` for these longer outputs?

Author: fixed.

```
---
## timescaledb_information.compression_settings [](timescaledb_information-compression_settings)

> Get information about compression related settings for hypertables.
> The view lists the orderby and segmentby columns used by compression.

Reviewer 2: ```suggestion
Get information about compression-related settings for hypertables.
Each row of the view provides information about individual orderby
and segmentby columns used by compression.
```

#### Available Columns [](timescaledb_information-compression_settings-available-columns)

|Name|Description|
|---|---|
| `table_schema` | Schema name of the hypertable |
| `table_name` | Table name of the hypertable |
| `attname` | Name of the column used in the compression settings |
| `segmentby_column_index` | Position of attname in the compress_segmentby list |
| `orderby_column_index` | Position of attname in the compress_orderby list |
| `orderby_asc` | (BOOLEAN) True if this is used for order by ASC, False for order by DESC |
| `orderby_nullsfirst` | (BOOLEAN) True if nulls are ordered first for this column, False if nulls are ordered last|

#### Sample Usage [](timescaledb_information-compression_settings-examples)

```sql

> CREATE TABLE hypertab (a_col integer, b_col integer, c_col integer, d_col integer);

Reviewer 2: Maybe include a column that _isn't_ in order/compress by, so user realizes that not all columns are listed in the below view?  (i.e., they might otherwise have false impression)

```suggestion
CREATE TABLE hypertab (a_col integer, b_col integer, c_col integer, d_col integer, e_col integer);
```

> SELECT table_name FROM create_hypertable('hypertab', 'a_col', chunk_time_interval=> 10);

Reviewer 2: Keep code examples as simple as possible ?
```suggestion
SELECT table_name FROM create_hypertable('hypertab', 'a_col');
```

|Name|Description|
|---|---|
|`hypertable` | (REGCLASS) The name of the hypertable on which the policy is applied |
|`job_id` | (INTEGER) The id of the background job created to implement the policy |
|`last_run_started_at`| Start time of the last job|
|`last_successful_finish`| Time when the job completed successfully|
|`last_run_status` | Whether the last run succeeded or failed |

> |`job_status`| Status of the job . Valid values are ‘Running’ and ‘Scheduled’|

Reviewer 2: ```suggestion
|`job_status`| Status of the job. Valid values are ‘Running’ and ‘Scheduled’|
```

|`last_run_duration`| Duration of last run of the job|
|`next_scheduled_run` | Start time of the next run |
|`total_runs` | The total number of runs of this job|
|`total_successes` | The total number of times this job succeeded |
|`total_failures` | The total number of times this job failed |

#### Sample Usage [](timescaledb_information-policy_stats-examples)

```
Get information about continuous aggregate policy related statistics

> ``` sql

Reviewer 2: Can we give a simpler example that doesn't involve a JOIN in addition?

Author: I added another example without a join.

SELECT  ps.* FROM
  timescaledb_information.policy_stats ps, timescaledb_information.continuous_aggregates cagg 

>   WHERE cagg.view_name = 'mat_m1'::regclass and cagg.materialization_hypertable = ps.hypertable;

Reviewer 2: Does this query need the ::regclass cast?  That type of thing is going to be more complicated for users to understand, so wonder if possible to do without a cast?

Author: Yes for now. @sven is making changes to the policy related views. This will most likely need another pass.

---

## hypertable_compression_stats() :community_function: [](hypertable_compression_stats)

Get statistics related to hypertable compression.
All sizes are in bytes.

#### Required Arguments [](hypertable_compression_stats-required-arguments)

|Name|Description|
|---|---|

> | `main_table` | (REGCLASS) Name of the hypertable.|

Reviewer 2: ```suggestion
| `main_table` | (REGCLASS) Name of the hypertable |
```

|`after_compression_table_bytes` | (BIGINT) Size of the heap after compression (NULL if currently uncompressed) |
|`after_compression_index_bytes` | (BIGINT) Size of all the indexes after compression (NULL if currently uncompressed) |
|`after_compression_toast_bytes` | (BIGINT) Size the TOAST table after compression (NULL if currently uncompressed) |
|`after_compression_total_bytes` | (BIGINT) Size of the entire table (table+indexes+toast) after compression (NULL if currently uncompressed) |
|`node_name` | (NAME) nodes on which the hypertable is located, applicable only to distributed hypertables |

#### Sample Usage [](hypertable_compression_stats-examples)
```sql
SELECT * FROM hypertable_compression_stats('conditions');

-[ RECORD 1 ]------------------+------
total_chunks                   | 4
number_compressed_chunks       | 1
before_compression_table_bytes | 8192
before_compression_index_bytes | 32768
before_compression_toast_bytes | 0
before_compression_total_bytes | 40960
after_compression_table_bytes  | 8192
after_compression_index_bytes  | 32768
after_compression_toast_bytes  | 8192
after_compression_total_bytes  | 49152
node_name                      |
```

> Use pg_size_pretty get the output in a more human friendly format.

Reviewer 2: ```suggestion
Use `pg_size_pretty` get the output in a more human friendly format.
```

```sql
SELECT pg_size_pretty(after_compression_total_bytes) as total

>   FROM hypertable_compression_stats('devices');

Reviewer 2: ```suggestion
  FROM hypertable_compression_stats('conditions');
```

-[ RECORD 1 ]--+------

> total | 32 kB

Reviewer 2: ```suggestion
total | 48 kB
```

```
---
## chunk_compression_stats() :community_function: [](chunk_compression_stats)

Get chunk specific statistics related to hypertable compression.
All sizes are in bytes.

#### Required Arguments [](chunk_compression_stats-required-arguments)

|Name|Description|
|---|---|

> | `main_table` | (REGCLASS) Name of the hypertable.|

Reviewer 2: ```suggestion
| `main_table` | (REGCLASS) Name of the hypertable |
```

#### Sample Usage [](chunk_compression_stats-examples)
```sql
SELECT * FROM chunk_compression_stats('conditions')
  ORDER BY chunk_name limit 2;

-[ RECORD 1 ]------------------+----------------------
chunk_schema                   | _timescaledb_internal
chunk_name                     | _hyper_1_1_chunk
compression_status             | Uncompressed
before_compression_table_bytes |
before_compression_index_bytes |
before_compression_toast_bytes |
before_compression_total_bytes |
after_compression_table_bytes  |
after_compression_index_bytes  |
after_compression_toast_bytes  |
after_compression_total_bytes  |
node_name                      |
-[ RECORD 2 ]------------------+----------------------
chunk_schema                   | _timescaledb_internal
chunk_name                     | _hyper_1_2_chunk
compression_status             | Compressed
before_compression_table_bytes | 8192
before_compression_index_bytes | 32768
before_compression_toast_bytes | 0
before_compression_total_bytes | 40960
after_compression_table_bytes  | 8192
after_compression_index_bytes  | 32768
after_compression_toast_bytes  | 8192
after_compression_total_bytes  | 49152
node_name                      |
```

> Use pg_size_pretty get the output in a more human friendly format.

Reviewer 2: ```suggestion
Use `pg_size_pretty` get the output in a more human friendly format.
```

```sql

> SELECT pg_size_pretty(after_compression_total_bytes) as total
>   FROM chunk_compression_stats('devices');
> 
> -[ RECORD 1 ]--+------
> total | 32 kB

Reviewer 2: ```suggestion
SELECT pg_size_pretty(after_compression_total_bytes) as total
  FROM chunk_compression_stats('conditions')
  WHERE compression_status = 'Compressed';

-[ RECORD 1 ]--+------
total | 48 kB
```
